### PR TITLE
Smarter reconciliation (take two)

### PR DESCRIPTION
**Pull request by hmac**
This PR implements the majority of https://github.com/gocardless/theatre/pull/21 on top of the latest changes in master. This was a lot more straightforward than attempting such a huge rebase.

Each commit should be a self-contained change, so hopefully it'll be easy to review. The one bit to look out for is 7fa8015, where some changes to the logic are made in an attempt to simplify it. I've not changed the tests at all (except for adding some additional expectations), so my hope is that their passing gives us confidence in the logic changes. If there was untested edge case handling in the previous code, we may have lost it.

### CODE REVIEW

File: pkg/workloads/console/controller.go [hunk 1]
```diff
@@ -74,115 +74,113 @@ func Add(ctx context.Context, logger kitlog.Logger, mgr manager.Manager, opts ..
...
+	if !isJobExpired(r.console) {
+		job = buildJob(r.name, consoleTemplate.Spec.Template)
+		if err := r.createOrUpdate(job, Job, jobDiff); err != nil {
 			return res, err
 		}
-	} else {
-		// Clear the 'not found' error, as we don't want this to cause a retry of
-		// the reconciliation.
-		err = nil
 		jobExists = true
```
Comment (hmac) [inline pkg/workloads/console/controller.go]: I really wanted to avoid having to keep track of this variable throughout reconciliation, but I couldn't see a clean way to do it on this pass. I think it's something we should revisit if we can.

**Comment by hmac** [comment]
@lawrencejones @benwh would love to get 👀 on this from both of you

File: pkg/reconcile/reconcile.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+// Operation describes the operation performed by CreateOrUpdate.
+type Operation string
+
+var (
+	Create   Operation = "create"
+	Recreate Operation = "recreate"
+	Update   Operation = "update"
+	None     Operation = "none"
+	Error    Operation = "error"
```
[the remark this answers is not in the document]
Comment (hmac) [reply pkg/reconcile/reconcile.go]: Yeah, I think the idea here is these describe what action the function took. Agree that "error" is not an action though. Result doesn't sound too bad to me. Or outcome?

File: pkg/reconcile/reconcile.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		return Update, nil
+	case None:
+		return None, nil
+	default:
+		return Error, fmt.Errorf("Unrecognised operation: %s", op)
+	}
+}
+
+// RoleDiff is a DiffFunc for Roles
+func RoleDiff(expectedObj runtime.Object, existingObj runtime.Object) Operation {
```
[the remark this answers is not in the document]
Comment (hmac) [reply pkg/reconcile/reconcile.go]: That's the idea. I think these ones are generic enough to be reused, but (for example) the job is specific to consoles because we ignore certain things we don't care about for the consoles use case.

File: pkg/workloads/console/controller.go [hunk 4]
```diff
@@ -193,52 +191,55 @@ func (r *ConsoleReconciler) Reconcile(request reconcile.Request) (res reconcile.
...
 
-	logger.Log("event", EventComplete)
+	r.logger.Log("event", EventComplete)
 	return res, err
 }
 
-func (r *ConsoleReconciler) updateRoleBindings(csl *workloadsv1alpha1.Console, tmpl *workloadsv1alpha1.ConsoleTemplate, name types.NamespacedName) error {
-	role := buildRole(name)
-	if err := controllerutil.SetControllerReference(csl, role, scheme.Scheme); err != nil {
+func (r *ConsoleReconciler) createOrUpdate(existing reconcile.ObjWithMeta, kind string, diffFunc reconcile.DiffFunc) error {
```
[the remark this answers is not in the document]
Comment (hmac) [reply pkg/workloads/console/controller.go]: In this case, yep! For `reconcile.CreateOrUpdate`, it's a bit more subtle. Will add a note there to explain.

File: pkg/reconcile/reconcile.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		if err := c.Create(ctx, existing); err != nil {
+			return Error, err
+		}
+		return Create, nil
+	}
+
+	// existing contains the state we just fetched from Kubernetes.
+	// expected contains the state we're trying to reconcile towards.
+	// If an update is required, DiffFunc will set the relevant fields on existing such that we
+	// can just resubmit it to the cluster to achieve our desired state.
```
Comment (hmac) [inline pkg/reconcile/reconcile.go]: The naming in this function is quite carefully chosen, and there's a subtle interaction between a couple of things here (all caused by pointers).

If you were to rename these variables such that we passed in an `expected` resource, you'd find that a load of tests fail. This is due to two things:
- We try to keep any existing state that Kube has set on the resource by fetching it, setting fields on it, and updating it
- The caller expects that, after calling `CreateOrUpdate`, the resource it passed it (as a pointer) will contain the most up-to-date state from Kube.

This means we must submit the exact resource passed in to Kube - not the copy we make for comparison purposes.

We could return the up-to-date resource from the function, and that would get around this, but I think we'd then need to type cast it back to the specific resource type, as the only type we could put in this function signature is `ObjectWithMeta`.

File: pkg/reconcile/reconcile.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	return None
+}
+
+// RoleBindingDiff is a DiffFunc for RoleBindings
+func RoleBindingDiff(expectedObj runtime.Object, existingObj runtime.Object) Operation {
+	expected := expectedObj.(*rbacv1.RoleBinding)
+	existing := existingObj.(*rbacv1.RoleBinding)
+	operation := None
+
+	if expected.ObjectMeta.Name != existing.ObjectMeta.Name {
```
[the remark this answers is not in the document]
Comment (hmac) [reply pkg/reconcile/reconcile.go]: This is true, in our case. In future though, I can imagine we might have controllers managing resources where the name differs, and in that case we'd want to do this. I'll add a comment explaining that we don't currently make use of this but might want to - I'm somewhat hesitant to remove it in case we forget to account for it in the future. What do you think?

File: pkg/reconcile/reconcile.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	Recreate Operation = "recreate"
+	Update   Operation = "update"
+	None     Operation = "none"
+	Error    Operation = "error"
+)
+
+// ObjWithMeta describes a Kubernetes resource with a metadata field. It's a combination
+// of two common existing Kubernetes interfaces. We do this because we want to use methods
+// from each in CreateOrUpdate, whilst still keeping the argument type generic.
+type ObjWithMeta interface {
```
Comment (lawrencejones) [inline pkg/reconcile/reconcile.go]: Any reason this is public? Do we expect anything outside of this package to make use of this interface?

File: pkg/reconcile/reconcile.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lawrencejones) [reply pkg/reconcile/reconcile.go]: I think we probably want to keep this, for a situation such as updating our code to produce slightly different resources: the names may change, so you'll want to recreate the resources.

This is thinking ahead of time but it's not too expensive for us to keep it now?

File: pkg/workloads/console/controller.go [hunk 8]
```diff
@@ -21,24 +21,24 @@ import (
...
-	EventComplete = "reconcile.end"
-	EventRequeued = "reconcile.requeued"
-	EventFound    = "found"
-	EventNotFound = "not_found"
-	EventCreated  = "created"
-	EventExpired  = "expired"
-	EventDeleted  = "deleted"
-	EventError    = "error"
+	EventStart    = "ReconcileStart"
+	EventComplete = "ReconcileEnd"
```
Comment (lawrencejones) [inline pkg/workloads/console/controller.go]: `ReconcileEnd` -> `ReconcileComplete`? Or maybe just drop the `Reconcile` prefix from them and keep them matching the constant name? I'd imagine having 121 mappings from constant name to the thing that appears in the logs would make debugging easier

File: pkg/workloads/console/controller.go [hunk 9]
```diff
@@ -74,115 +74,113 @@ func Add(ctx context.Context, logger kitlog.Logger, mgr manager.Manager, opts ..
 	return ctrl, err
 }
 
-type ConsoleReconciler struct {
+type Controller struct {
```
[the remark this answers is not in the document]
Comment (lawrencejones) [reply pkg/workloads/console/controller.go]: I'd expect us to be placing this specialised type of controller, where we want to retrieve our resource and construct a recording logger, along with installing an error handler, to live in the reconcile package we've created and wrap a generic reconciler. Reckon it's pretty valuable that everything initialises in a consistent fashion!

File: pkg/workloads/console/controller.go [hunk 10]
```diff
@@ -74,115 +74,113 @@ func Add(ctx context.Context, logger kitlog.Logger, mgr manager.Manager, opts ..
...
 	client   client.Client
 }
 
-func (r *ConsoleReconciler) Reconcile(request reconcile.Request) (res reconcile.Result, err error) {
-	logger := kitlog.With(r.logger, "request", request)
+func (c *Controller) Reconcile(request k8rec.Request) (k8rec.Result, error) {
+	logger := kitlog.With(c.logger, "request", request)
 	logger.Log("event", EventStart)
 
-	defer func() {
```
Comment (lawrencejones) [inline pkg/workloads/console/controller.go]: How come no defer? I was in favour of the defer to ensure we always log the error, and could omit the logging from any place that generated one.

File: pkg/workloads/console/controller.go [hunk 11]
```diff
@@ -74,115 +74,113 @@ func Add(ctx context.Context, logger kitlog.Logger, mgr manager.Manager, opts ..
...
 		if errors.IsNotFound(err) {
-			return res, logger.Log("event", EventNotFound, "resource", Console)
+			logger.Log("event", EventNotFound, "resource", Console)
 		}
-		return res, err
+		return k8rec.Result{}, err
 	}
 
-	logger = logging.WithRecorder(logger, r.recorder, csl)
+	// This is temporarily disabled as our logs don't pass k8s event validation
```
Comment (lawrencejones) [inline pkg/workloads/console/controller.go]: Are they still not passing?

File: pkg/workloads/console/controller.go [hunk 12]
```diff
@@ -74,115 +74,113 @@ func Add(ctx context.Context, logger kitlog.Logger, mgr manager.Manager, opts ..
...
+		console: csl,
+	}
+	result, err := reconciler.Reconcile()
+	if err != nil {
+		logger.Log("event", EventError, "error", err)
+	}
+	return result, err
+}
+
+type ConsoleReconciler struct {
```
[the remark this answers is not in the document]
Comment (lawrencejones) [reply pkg/workloads/console/controller.go]: Given we've called `Controller` as such, I think dropping the prefix makes sense.

Question here about making this private. What are people's thoughts on that?

File: pkg/workloads/console/controller.go [hunk 13]
```diff
@@ -360,3 +330,22 @@ func calculateStatus(csl *workloadsv1alpha1.Console, job *batchv1.Job) workloads
...
+// jobDiff is a reconcile.DiffFunc for Jobs
+func jobDiff(expectedObj runtime.Object, existingObj runtime.Object) reconcile.Operation {
+	expected := expectedObj.(*batchv1.Job)
+	existing := existingObj.(*batchv1.Job)
+	if expected.ObjectMeta.Name != existing.ObjectMeta.Name {
+		return reconcile.Recreate
+	}
+
+	if !reflect.DeepEqual(expected.Spec.Template, existing.Spec.Template) {
+		// We don't update the pod template's metadata, as this has already been modified by
```
[the remark this answers is not in the document]
Comment (hmac) [reply pkg/workloads/console/controller.go]: Can't remember off the top of my head - will investigate and get back to you

[the remark this answers is not in the document]
Comment (hmac) [reply pkg/workloads/console/controller.go]:
```
{
    Type: "FieldValueRequired",
    Message: "Required value: must be '1c97f196-23b1-11e9-993d-acde48001122'",
    Field: "spec.template.metadata.labels[controller-uid]",
},
{
    Type: "FieldValueRequired",
    Message: "Required value: must be 'console-0'",
    Field: "spec.template.metadata.labels[job-name]",
},
{
    Type: "FieldValueInvalid",
    Message: "Invalid value: map[string]string(nil): `selector` does not match template `labels`",
    Field: "spec.template.metadata.labels",
},
{
    Type: "FieldValueInvalid",
    Message: "Invalid value: map[string]string(nil): `selector` does not match template `labels`",
    Field: "spec.template.metadata.labels",
}
```

It looks like these fields get set by Kube when we create the resource.

File: pkg/workloads/console/controller.go (hunk 8 above)
Comment (hmac) [reply pkg/workloads/console/controller.go]: thought you were saying it would be good to have one hundred and twenty-one mappings for a minute..

File: pkg/workloads/console/controller.go (hunk 10 above)
Comment (hmac) [reply pkg/workloads/console/controller.go]: We now log the error in the outer `Controller` - the `Reconciler` just returns the error. Same result (I think) - slightly cleaner?

File: pkg/workloads/console/controller.go (hunk 11 above)
Comment (hmac) [reply pkg/workloads/console/controller.go]: no dice

File: pkg/reconcile/reconcile.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (benwh) [reply pkg/reconcile/reconcile.go]: Outcome fits well IMO!

File: pkg/workloads/console/controller.go (hunk 10 above)
Comment (lawrencejones) [reply pkg/workloads/console/controller.go]: Ah gotcha, my mistake. 👍

File: pkg/workloads/console/controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (benwh) [reply pkg/workloads/console/controller.go]: Ah, this is obvious in hindsight. I think maybe we need to fix up the comment (just `job's metadata`?) to reduce confusion though

File: pkg/workloads/console/controller.go [hunk 14]
```diff
@@ -234,11 +234,12 @@ func (r *ConsoleReconciler) updateRoleBindings(tmpl *workloadsv1alpha1.ConsoleTe
...
 		return err
 	}
-	if err := findOrCreate(r.ctx, r.client, role, r.name); err != nil {
+	operation, err := reconcile.CreateOrUpdate(r.ctx, r.client, role, Role, reconcile.RoleDiff)
+	if err != nil {
 		r.logger.Log("event", EventError, "resource", Role, "error", err)
 		return err
 	}
-	r.logger.Log("event", EventCreated, "resource", Role)
+	r.logger.Log("event", operation, "resource", Role)
```
[the remark this answers is not in the document]
Comment (hmac) [reply pkg/workloads/console/controller.go]: Have nested this value under an `outcome` key.

**Comment by hmac** [comment]
Tests for the bug that Ben pointed out are probably best addressed in the acceptance flow. Have added an issue for that here: https://github.com/gocardless/theatre/issues/24

File: pkg/workloads/console/integration/integration_test.go [hunk 15]
```diff
@@ -264,14 +271,12 @@ var _ = Describe("Console", func() {
...
-			job := &batchv1.Job{}
+			job = &batchv1.Job{}
 			identifier, _ = client.ObjectKeyFromObject(csl)
 			err = mgr.GetClient().Get(context.TODO(), identifier, job)
 
-			statusError, _ := err.(*errors.StatusError)
 			Expect(err).To(HaveOccurred(), "the job should remain deleted, but it has been recreated")
-			Expect(err).To(BeAssignableToTypeOf(statusError), "error should be an API error")
-			Expect(statusError.Status()).To(MatchFields(IgnoreExtras, Fields{"Code": BeEquivalentTo(404)}))
+			Expect(errors.IsNotFound(err)).To(Equal(true), "a NotFound error was expected, but got a different error")
```
Comment (benwh) [inline pkg/workloads/console/integration/integration_test.go]: Minor note: The downside of this approach is that if the assertion fails then the error message is not going to be of much use (i.e. `true != false`); we'd have to insert some extra logging to find out what went wrong

Comment (hmac) [reply pkg/workloads/console/integration/integration_test.go]: That's true, but I'd argue that the tests become easier to read as a result. I guess it's a balance of whether you're optimising for reading or debugging.
